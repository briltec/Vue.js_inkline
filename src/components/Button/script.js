import AttributesProviderMixin from 'inkline/mixins/components/providers/AttributesProviderMixin';
import ClassesProviderMixin from 'inkline/mixins/components/providers/ClassesProviderMixin';
import InjectParentFormProviderMixin from 'inkline/mixins/forms/providers/InjectParentFormProviderMixin';

import ClickMethodMixin from 'inkline/mixins/forms/methods/ClickMethodMixin';
import FocusMethodMixin from 'inkline/mixins/forms/methods/FocusMethodMixin';
import OnClickMethodMixin from 'inkline/mixins/components/methods/OnClickMethodMixin';
import OnFocusMethodMixin from 'inkline/mixins/components/methods/OnFocusMethodMixin';
import OnHoverMethodMixin from 'inkline/mixins/components/methods/OnHoverMethodMixin';

import DisabledPropertyMixin from 'inkline/mixins/forms/properties/DisabledPropertyMixin';
import LoadingPropertyMixin from 'inkline/mixins/components/properties/LoadingPropertyMixin';
import SizePropertyMixin from 'inkline/mixins/components/properties/SizePropertyMixin';
import TabIndexPropertyMixin from 'inkline/mixins/components/properties/TabIndexPropertyMixin';
import VariantPropertyMixin from 'inkline/mixins/components/properties/VariantPropertyMixin';

export default {
    name: 'IButton',
    mixins: [
        AttributesProviderMixin,
        ClassesProviderMixin,
        InjectParentFormProviderMixin,

        ClickMethodMixin,
        FocusMethodMixin,
        OnClickMethodMixin,
        OnFocusMethodMixin,
        OnHoverMethodMixin,

        DisabledPropertyMixin,
        LoadingPropertyMixin,
        SizePropertyMixin,
        TabIndexPropertyMixin,
        VariantPropertyMixin
    ],
    props: {
        /**
         * Modifiers
         */
        active: {
            type: Boolean,
            default: false
        },
        block: {
            type: Boolean,
            default: false
        },
        circle: {
            type: Boolean,
            default: false
        },
        flat: {
            type: Boolean,
            default: false
        },
        link: {
            type: Boolean,
            default: false
        },
        outline: {
            type: Boolean,
            default: false
        },
        tag: {
            type: String,
            default: 'button'
        },

        /**
         * Icon
         */
        icon: {
            type: String,
            default: ''
        },
    },
    computed: {
        isTag () {
            return this.attributes.to ? 'router-link' : this.attributes.href ? 'a' : this.tag;
        }
    },
    created () {
        if (this.classesProvider) {
            this.classesProvider['root'].push(() => ({'-active': this.active}));
            this.classesProvider['root'].push(() => ({'-block': this.block}));
            this.classesProvider['root'].push(() => ({'-circle': this.circle}));
            this.classesProvider['root'].push(() => ({'-flat': this.flat}));
            this.classesProvider['root'].push(() => ({'-link': this.link}));
            this.classesProvider['root'].push(() => ({'-outline': this.outline}));
        }

        if (this.attributesProvider) {
            this.attributesProvider.push(() => ({ 'aria-pressed': this.active ? 'true' : false }));
        }
    }
};
