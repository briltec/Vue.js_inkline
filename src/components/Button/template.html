<button
    v-bind="attributes"
    class="button"
    :is="isTag"
    :tag="tag"
    :tabindex="tabIndex"
    :class="classes"
    :disabled="isDisabled || loading"
    @click="emitClick"
>
    <slot v-if="loading" name="loading">Loading..</slot>
    <slot v-if="!loading"></slot>
</button>
