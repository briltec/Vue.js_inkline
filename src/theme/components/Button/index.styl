//
// Variables
//
--button-themes := (basic brand neutral social state)

--button-box-shadow := --box-shadow[0] / 2 0.25rem 0.25rem -0.25rem rgba(--colors['black'], 0.33)
--button-box-shadow-focus := --box-shadow[0] 0.5rem 0.75rem -0.5rem rgba(--colors['black'], 0.66)
--button-box-shadow-active := --box-shadow[0] 0 0 0 rgba(--colors['black'], 0.33)
--button-box-shadow-active-inset := inset 0 0 0.5rem rgba(--colors['black'], 0.1)

--button-darken-percentage-hover := 5%

//
// Themes
//
-add-theme(--themes, 'button', --button-themes)

//
// Styles
//
.button
    box-shadow: --button-box-shadow
    background-repeat: repeat
    background-position: 0 0
    background-size: 100% 100%

    if --transition['button']
        transition: background-color --transition-duration --transition-easing, background-position --transition-duration --transition-easing, background-size --transition-duration --transition-easing, border-color --transition-duration --transition-easing, color --transition-duration --transition-easing, box-shadow --transition-duration --transition-easing !important

    &:hover
    &.-hovered
    &:focus
    &.-focused
        background-size: 150% 150%
        background-position: 100% 100%
        box-shadow: --button-box-shadow-focus

    // Active state
    &:active
    &.-active
        background-size: 200% 200%
        background-position: 100% 100%
        box-shadow: --button-box-shadow-active, --button-box-shadow-active-inset

    // Link buttons
    // Make a button look and behave like a link
    &.-link
        box-shadow: none

        &:not(:disabled):not(.-disabled)
            &:hover
            &.-hovered
            &:focus
            &.-focused
                box-shadow: none

    &.-outline
        background-color: transparent
        background-position: 0 0
        background-size: 300% 300%
        box-shadow: none

        &:not(:disabled):not(.-disabled)
            &:hover
            &.-hovered
            &:focus
            &.-focused
                background-size: 300% 300%
                background-position: 100% 100%
                box-shadow: --button-box-shadow-focus

            &:active
            &.-active
                background-size: 400% 400%
                background-position: 100% 100%
                box-shadow: --button-box-shadow-active, --button-box-shadow-active-inset

    // Button themes
    // Add the required color themes to all button variations
    for --theme-name, --theme in --themes['button']
        &.-{--theme-name}
            color: -text-color-by-luminance(--theme['color'], --theme-color-light, --theme-color-dark)
            background-color: --theme['color']
            border-color: --theme['color']

            if --gradient-enabled and --theme['gradient']
                background-image: -linear-gradient(--theme['gradient'])

            &:not(:disabled):not(.-disabled)
                &:hover
                &:focus
                &.-hovered
                &.-focused
                    background-color: darken(--theme['color'], --button-darken-percentage-hover)
                    border-color: darken(--theme['color'], --button-darken-percentage-hover)

                &.:active
                &.-active
                &:focus
                &.-focused
                    outline: 1px solid rgba(--theme['color'], 0.25)
                    outline-offset: 2px

            &.-outline
                background-color: transparent
                color: --theme['color']
                background-image: -linear-gradient(transparent transparent (--theme['gradient'][0]) (--theme['gradient'][1]))

                &:not(:disabled):not(.-disabled)
                    &:hover
                    &.-hovered
                    &:focus
                    &.-focused
                        color: -text-color-by-luminance(--theme['color'], --theme-color-light, --theme-color-dark)
                        background-color: darken(--theme['color'], --button-darken-percentage-hover)
                        border-color: darken(--theme['color'], --button-darken-percentage-hover)

            &.-link
                color: --theme['color']
                background-color: transparent
                background-image: none
                border-color: transparent

                &:not(:disabled):not(.-disabled)
                    &:hover
                    &.-hovered
                    &:focus
                    &.-focused
                        color: darken(--theme['color'], 5%)
                        background-color: transparent
                        background-image: none
                        border-color: transparent

// Button defaults
// Set a default size and theme
.button
    @extend .button.-light