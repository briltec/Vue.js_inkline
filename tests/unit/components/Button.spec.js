import { mount } from '@vue/test-utils';
import Button from 'inkline/components/Button';

describe('Components', () => {
    describe('Button', () => {
        let wrapper;

        beforeEach(() => {
            wrapper = mount(Button);
        });

        test('should render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });
    });
});
