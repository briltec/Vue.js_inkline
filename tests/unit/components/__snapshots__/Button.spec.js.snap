// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Button should render correctly 1`] = `<div role="checkboxgroup" class="button-group form-group -checkbox"></div>`;
